lib/corefoundation/register.rb

```diff
@@ -0,0 +1,32 @@
+require "singleton"
+
+module CF
+  module Register
+    class RegisteredTypes
```

Comment:
Can we have this as a class object or frozen hash instead of module?
We can then use that in `base` by initializing and setting it as a class variable.
You can still have `private_constant` call separately, though I would discourage from that, since the constant could be re-defined/overriden either way.

Making `type_map` immutable is a good thought, overall